### DISCUSSION TITLE
difference

### DISCUSSION THREAD

**Discussion by Oops-Qiao** [idaholab/moose#22249] [Q&A General] [answered]
Hello,

There is a question in the kernel,
Does the following two play the same role in the kernel? Do they have difference or not? Thank you.

ADTimeDerivative::precomputeQpResidual();  
_u_dot[_qp]

**Comment by GiudGiud** [comment, marked as answer]
Hello

The are the same
```
ADReal
ADTimeDerivative::precomputeQpResidual()
{
  return _u_dot[_qp];
}
```

However, this is a precomputeQpResidual routine. So it must be defined alongside a computeQpResidual routine that uses it. 
Be careful with this, it's an optimization and if you dont use it properly you will miss the test function multiplication

Guillaume

**Reply by Oops-Qiao** [reply]
Ok， thank you Guillaume, one question, here
ADTimeDerivative::precomputeQpResidual() is defined as ADReal type. Can I still use it if I have vector like velocity?

**Reply by GiudGiud** [reply]
No you ll have to use the corresponding type for an ADvectorKernel. 
It is likely ADRealVectorValue

**Reply by Oops-Qiao** [reply]
> No you ll have to use the corresponding type for an ADvectorKernel. It is likely ADRealVectorValue

So it is better use  _u_dot[_qp] in the kernel instead of ADTimeDerivative::precomputeQpResidual(), right?

**Reply by GiudGiud** [reply]
Yes but don’t forget the test function that’s it

**Reply by Oops-Qiao** [reply]
> Yes but don’t forget the test function that’s it



> Yes but don’t forget the test function that’s it

Sure, since I inherit ADKernelValue, test function will be automatically applied.

**Reply by Oops-Qiao** [reply]
> Hello
> 
> The are the same
> 
> ```
> ADReal
> ADTimeDerivative::precomputeQpResidual()
> {
>   return _u_dot[_qp];
> }
> ```
> 
> However, this is a precomputeQpResidual routine. So it must be defined alongside a computeQpResidual routine that uses it. Be careful with this, it's an optimization and if you dont use it properly you will miss the test function multiplication
> 
> Guillaume


Hi Guillaume,

Today just a question came to my mind regarding your answer above.

For example the code below in your moose,
> ADReal
> ADTimeDerivative::precomputeQpResidual()
> {
>   return _u_dot[_qp];

[ADHeatConductionTimeDerivative.C
](https://github.com/idaholab/moose/blob/next/modules/heat_conduction/src/kernels/ADHeatConductionTimeDerivative.C)
ADReal
ADHeatConductionTimeDerivative::**precomputeQpResidual()**
{
  return _specific_heat[_qp] * _density[_qp] * ADTimeDerivative::**precomputeQpResidual()**;
}

I am wondering that if the two bold precomputeQpResidual() plays the same role to multiply the test function, if that is the case, it looks a little bit messed up since we only need to multply once the test function. 
What do you suggest to understand it?
Thank you so much.

Qiao

**Reply by maxnezdyur** [reply]
This has to do with you modifying precomputeQPResidual. If using precomputeQPResidual then you do not multiply the test function. The link explains it more in-depth https://mooseframework.inl.gov/magpie/syntax/Kernels/ under **Optimized Kernel Objects**.

**Reply by Oops-Qiao** [reply]
> This has to do with you modifying precomputeQPResidual. If using precomputeQPResidual then you do not multiply the test function. The link explains it more in-depth https://mooseframework.inl.gov/magpie/syntax/Kernels/ under **Optimized Kernel Objects**.

Thanks, that is also my understanding. However, if you see below, it has two times of precomputeQPResidual which means it will then multiply the test function two times. Is it right? This makes me confused.  
@GiudGiud 

 ADReal
ADHeatConductionTimeDerivative::precomputeQpResidual()
{
return _specific_heat[_qp] * _density[_qp] * ADTimeDerivative::precomputeQpResidual();
}

**Reply by maxnezdyur** [reply]
I don't think so. Only the "final" **precomputeQpResidual()** gets multiplied by the test functions. IIRC this is an example of overriding a virtual function so only the ADHeatConductionTimeDerivative::precomputeQpResidual() gets called and multiplied by the test functions. Even though ADTimeDerivative::precomputeQpResidual() is called also by ADHeatConductionTimeDerivative , only the _u_dot[_qp] is incorporated with the ADHeatConductionTimeDerivative. This explanation may be confusing, so sorry for that!

**Reply by Oops-Qiao** [reply]
> ADReal
> ADTimeDerivative::precomputeQpResidual()
> {
> return _u_dot[_qp];

I really appreciate your patient reply. Just one more question here. if you say “only the _u_dot[_qp] is incorporated ” . 
If we now take look at the ADTimeDerivative document, they have used the following definition with  "precomputeQpResidual()" in below.
Do we have test function multiplied in the code below? If Yes, how to understand your current explanation then. Thanks again.

ADReal
ADTimeDerivative::precomputeQpResidual()
{
return _u_dot[_qp];
}

**Reply by maxnezdyur** [reply]
I recommend looking at ADKernelValue.C on line 43 and 45. You will see **precomputeQpResidual()** being called and on line 45 multiplied by the test functions. This is when it happens in the code. Therefore it's the only time it get's multiplied by the test functions.
